### ISSUE TITLE
Tensor.nonzero fails on GPU for tensors containing more than INT_MAX elements

### ISSUE THREAD

**Issue by mberr** [pytorch/pytorch#51871]
## 🐛 Bug

Calling nonzero for a tensor with more than INT_MAX elements leads to an error on GPU. On CPU there does not seem to be a problem.

## To Reproduce

```python-console
>>> x = torch.zeros(2**20, 2**11, device="cuda")
>>> x.nonzero()
Traceback (most recent call last):
  File "<input>", line 1, in <module>
RuntimeError: nonzero is not supported for tensors with more than INT_MAX elements,   file a support request
```
## Expected behavior

Return the indices of non-zero elements.

## Environment

```
Collecting environment information...
PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti

Nvidia driver version: 450.102.04
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect
```

## Additional context

The problem was already described on [discuss.pytorch.org](https://discuss.pytorch.org/t/torch-geometric/106975), but I could not find the corresponding issue.

cc @ngimel

**Comment by RylanSchaeffer** [comment]
Has any progress been made on this error? I'm currently hitting it.

**Comment by kalikhademi** [comment]
I have the same issue when I am using pytorch with detectron 2.

**Comment by marcus-r-kelly** [comment]
I also am having this issue. Additionally, is there some way that I can find out the value of INT_MAX?

**Comment by bhack** [comment]
@ezyang What do you think about this? Is there a workaround in lowering this with compile? If you see we have many open tickets related this limit.

**Comment by ezyang** [comment]
I took a quick look at the implementation to see how hard it would be to fix, but unfortunately we use CUB to implement the bulk of the algorithm, and the algo with use Flagged https://github.com/NVIDIA/cccl/blob/main/cub/cub/device/device_select.cuh does not support int64_t numel.

One userland workaround is to partition the input into chunks that fit in 32-bit, run nonzero on them individually, and cat the results together. We could also try rewriting the algorithm to thrust.

I'm happy to review someone who sends in code, but I probably am not going to get around to it myself.

**Comment by bhack** [comment]
Is this going to impact also boolean masking ops right? E.g `mytensor[mask]`

**Comment by bhack** [comment]
I could be interested but:
- As this is at c++/cuda level do we have a nightly cache to speedup the c++ build? if not I suppose it will take forever to compile.
- What is the minimal test to be invoked locally for an initial check of an alternative implementation?

**Comment by bhack** [comment]
In case anyone is interested in the upstream ticket  https://github.com/NVIDIA/cccl/issues/47

**Comment by bhack** [comment]
See also https://github.com/NVIDIA/cccl/issues/50

**Comment by bhack** [comment]
@ezyang Is this enough when merged/released :
https://github.com/NVIDIA/cccl/issues/1422

**Comment by ezyang** [comment]
Yeah, it would be very simple to do the rest.

Re building Pytorch, you are just going to have to suck it up and do a cuda build from scratch, no cloud cache alas

**Comment by bhack** [comment]
@ezyang I don't know the ETA for the upstream merge but what do you think about the @elstehle temp wrapper at 
https://github.com/NVIDIA/cccl/pull/1612 ?

**Comment by bhack** [comment]
I think it could be still risky to temp adapt a solution like that wrapper:

> We want to have more elaborate test coverage by adding tests for large number of items to the CUB tests of cub::DeviceSelect, since we have better testing infrastructure in CUB.

**Comment by ezyang** [comment]
This is less a big deal for nonzero, which would be exercising CUB in one particular way, so we'd mostly just need a decent number of tests at large size for nonzero.

**Comment by bhack** [comment]
This is going to happen also in inductor emitted code. 
E.g.
```python
aten.index_put_(buf4, [reinterpret_tensor(buf0, (1, 1, s2*s3, 14 + s2, 14 + s3), (0, 0, 196 + (14*s2) + (14*s3) + (s2*s3), 14 + s3, 1), 0)], buf6, False)
```
```python
RuntimeError: nonzero is not supported for tensors with more than INT_MAX elements,   See https://github.com/pytorch/pytorch/issues/51871
~~~~

**Comment by surya-narayanan** [comment]
Also experienced this issue when I was trying to load big images into segment-anything.

**Comment by bhack** [comment]
Also the just released segment-anything v2 https://github.com/facebookresearch/segment-anything-2/issues/44

**Comment by jrhemstad** [comment]
fyi, the `cub::DeviceSelect` algorithm will be updated to support more than `INT_MAX` input elements when this PR is merged (nearly completed): https://github.com/NVIDIA/cccl/pull/2400

**Comment by bhack** [comment]
> fyi, the `cub::DeviceSelect` algorithm will be updated to support more than `INT_MAX` input elements when this PR is merged (nearly completed): [NVIDIA/cccl#2400](https://github.com/NVIDIA/cccl/pull/2400)

Thank you for the  update. Do you already know the target CUDA release?

**Comment by galv** [comment]
@bhack you can depend upon a version of cub orthogonal to the cub version released with a particular version of the cuda toolkit by adding a dependency to cccl in cmake: https://github.com/NVIDIA/cccl/?tab=readme-ov-file#cmake-integration I suppose it might be tedious to make it work for the bazel/buck build in pytorch, but I am surprised that pytorch appears to depend upon the cub version in the cuda toolkit today (https://github.com/pytorch/pytorch/blob/main/cmake/Modules/FindCUB.cmake), to be honest.

**Comment by galv** [comment]
If the reason for removing third_party/cub was that it was considered stable and unchanging, then this could be a good reasons to bring back third_party/cub as third_party/cccl then.

**Comment by bhack** [comment]
> If the reason for removing third_party/cub was that it was considered stable and unchanging, then this could be a good reasons to bring back third_party/cub as third_party/cccl then.

I don't know the original rationale. @maffet reviewed this 3 years ago at https://github.com/pytorch/pytorch/pull/62445

**Comment by bhack** [comment]
Or probably it is older coming from caffe2 https://github.com/facebookarchive/caffe2/pull/382

**Comment by jing-4369** [comment]
Has this problem been resolved？
or is there a easy way by updating CUDAtoolkit or something?

**Comment by bhack** [comment]
Upstream (cccl) was solved with https://github.com/NVIDIA/cccl/pull/2400

We need to wait a new cccl release, adopt this release in pytorch when it is going to be released in a new cuda version (https://github.com/pytorch/pytorch/issues/125981#issuecomment-2110734485)  and then make a pytorch PR.

**Comment by cpuhrsch** [comment]
@bhack - We're adding support for 12.6: https://github.com/pytorch/pytorch/issues/138609

**Comment by bhack** [comment]
> @bhack - We're adding support for 12.6: https://github.com/pytorch/pytorch/issues/138609

But this will require CCCL 2.8 https://github.com/NVIDIA/cccl/pull/2400#issuecomment-2494987280

**Comment by zzhuolun** [comment]
I solved it by upgrading torch (`2.5.1`-> `2.6.0`)
